packages/engine-core/src/library/LibraryEngine.ts

```diff
@@ -320,63 +322,87 @@ You may have to run ${chalk.greenBright('prisma generate')} for your changes to
-        this.engine
-          ?.connect({ enableRawQueries: true })
-          .then(() => {
-            this.libraryStarted = true
-            debug('library started')
-            resolve()
-          })
-          .catch((err) => {
-            const error = this.parseInitError(err.message)
-            // The error message thrown by the query engine should be a stringified JSON
-            // if parsing fails then we just reject the error
-            if (typeof error === 'string') {
-              reject(err)
-            } else {
-              reject(new PrismaClientInitializationError(error.message, this.config.clientVersion!, error.error_code))
-            }
-          })
-          .finally(() => {
-            this.libraryStartingPromise = undefined
-          })
-      })
-      return this.libraryStartingPromise
+
+    if (this.libraryStarted) {
+      return
+    }
+
+    const tracingConfig = getTracingConfig(this)
+
+    const startFn = async () => {
+      debug('library starting')
+
+      try {
+        await this.engine?.connect(getTraceParent())
+
+        this.libraryStarted = true
+
+        debug('library started')
+      } catch (err) {
+        const error = this.parseInitError(err.message as string)
+
+        // The error message thrown by the query engine should be a stringified JSON
+        // if parsing fails then we just reject the error
+        if (typeof error === 'string') {
+          throw err
+        } else {
+          throw new PrismaClientInitializationError(error.message, this.config.clientVersion!, error.error_code)
+        }
+      } finally {
+        this.libraryStartingPromise = undefined
+      }
+    }
+
+    if (tracingConfig.enabled) {
+      this.libraryStartingPromise = runInActiveSpan({ name: 'prisma:connect', callback: startFn })
+    } else {
+      this.libraryStartingPromise = startFn()
     }
+
+    return this.libraryStartingPromise
   }
 
   async stop(): Promise<void> {
     await this.libraryStartingPromise
     await this.executingQueryPromise
+
     if (this.libraryStoppingPromise) {
       debug('library is already stopping')
       return this.libraryStoppingPromise
     }
 
-    if (this.libraryStarted) {
-      // eslint-disable-next-line no-async-promise-executor
-      this.libraryStoppingPromise = new Promise(async (resolve, reject) => {
-        try {
-          await new Promise((r) => setTimeout(r, 5))
-          debug('library stopping')
-          await this.engine?.disconnect()
-          this.libraryStarted = false
-          this.libraryStoppingPromise = undefined
-          debug('library stopped')
-          resolve()
-        } catch (err) {
-          reject(err)
-        }
-      })
-      return this.libraryStoppingPromise
+    if (!this.libraryStarted) {
+      return
     }
+
+    const tracingConfig = getTracingConfig(this)
+
+    const stopFn = async () => {
+      // TODO - What is doing, why '5'?
+      await new Promise((r) => setTimeout(r, 5))
```

Comment:
I think this was to avoid a race condition in the event loop, probably something isn't awaited properly somewhere.

Reply:
Could be because of the use of `async` in `Promise`, which is an anti-pattern. So we can maybe just remove this line.

Reply:
I have removed it! 🎉

Reply:
Adding this back because it makes this test fail: https://github.com/prisma/prisma/blob/main/packages/client/src/__tests__/integration/happy/disconnect-race/test.ts